addons/l10n_es_edi_tbai/models/account_edi_format.py

```diff
@@ -0,0 +1,412 @@
+        )):
+            values['regime_key'].append(51)  # code for invoices with "recargo" taxes
+
+        return values
+
+    def _l10n_es_tbai_get_trail_values(self, invoice, cancel):
+        prev_invoice = invoice.company_id._get_l10n_es_tbai_last_posted_invoice(invoice)
+        # NOTE: assumtion that last posted == previous only works because XML is generated on post
+        # (the edi_content will change when another invoice is posted)
+        if prev_invoice and not cancel:
+            return {
+                'chain_prev_invoice': prev_invoice
+            }
+        else:
+            return {}
+
+    def _l10n_es_tbai_sign_invoice(self, invoice, xml_root):
+        company = invoice.company_id
+        cert_private, cert_public = company.l10n_es_edi_certificate_id._get_key_pair()
+        public_key = cert_public.public_key()
+
+        # Identifiers
+        document_id = "Document-" + str(uuid4())
+        signature_id = "Signature-" + document_id
+        keyinfo_id = "KeyInfo-" + document_id
+        sigproperties_id = "SignatureProperties-" + document_id
+
+        # Render digital signature scaffold from QWeb
+        web_services = get_web_services_for_agency(agency=company.l10n_es_tbai_tax_agency, env=self.env)
+        common_name = cert_public.issuer.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value
+        org_unit = cert_public.issuer.get_attributes_for_oid(NameOID.ORGANIZATIONAL_UNIT_NAME)[0].value
+        org_name = cert_public.issuer.get_attributes_for_oid(NameOID.ORGANIZATION_NAME)[0].value
+        country_name = cert_public.issuer.get_attributes_for_oid(NameOID.COUNTRY_NAME)[0].value
+        values = {
+            'dsig': {
+                'document_id': document_id,
+                'x509_certificate': bytes_as_block(b64encode(cert_public.public_bytes(encoding=serialization.Encoding.DER))),
+                'public_modulus': bytes_as_block(int_as_bytes(public_key.public_numbers().n)),
+                'public_exponent': bytes_as_block(int_as_bytes(public_key.public_numbers().e)),
+                'iso_now': datetime.now().isoformat(),
+                'keyinfo_id': keyinfo_id,
+                'signature_id': signature_id,
+                'sigproperties_id': sigproperties_id,
+                'reference_uri': "Reference-" + document_id,
+                'sigpolicy_description': "Política de Firma TicketBAI 1.0",  # í = &#237;
+                'sigpolicy_url': web_services.get_sigpolicy_url(),
+                'sigpolicy_digest': web_services.get_sigpolicy_digest(),
+                'sigcertif_digest': b64encode(cert_public.fingerprint(hashes.SHA256())).decode(),
+                'x509_issuer_description': 'CN={}, OU={}, O={}, C={}'.format(common_name, org_unit, org_name, country_name),
+                'x509_serial_number': cert_public.serial_number,
+            }
+        }
+        xml_sig_str = self.env['ir.qweb']._render(self.env.ref('l10n_es_edi_tbai.template_digital_signature').id, values)
```

Comment:
Are you sure we can not just pass the name of the template immediately?

Reply:
We can.